File: Dsa-day83.py
from bisect import bisect_left

class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        potions.sort()
        n = len(spells)
        m = len(potions)
        pairs = [0] * n

        for i in range(n):
            spell_strength = spells[i]
            
            required_potion_strength = (success + spell_strength - 1) // spell_strength
            
            k = bisect_left(potions, required_potion_strength)
            
            pairs[i] = m - k
            
        return pairs

File: Dsa-day83.java
import java.util.Arrays;

class Solution {
    public int[] successfulPairs(int[] spells, int[] potions, long success) {
        Arrays.sort(potions);
        
        int n = spells.length;
        int m = potions.length;
        int[] pairs = new int[n];

        for (int i = 0; i < n; ++i) {
            long spell_strength = spells[i];
            
            long required_potion_strength = (success + spell_strength - 1) / spell_strength;
            
            int low = 0;
            int high = m;
            int k = m; 

            while (low < high) {
                int mid = low + (high - low) / 2;
                if (potions[mid] < required_potion_strength) {
                    low = mid + 1;
                } else {
                    k = mid;
                    high = mid;
                }
            }
            
            pairs[i] = m - k;
        }
        
        return pairs;
    }
}